File: src/lexer/lexer.erl
-module(lexer).
-export([init/1, next_token/1]).

-record(lexer, {position = 0, read_position = 0, ch = 0, input = []}).

%% Types
-type token() :: ident | integer | illegal | eof |
                equal | plus | comma | semicolon |
                lparen | rparen | lsquirly | rsquirly |
                function | let.

init(Input) ->
    Lexer = #lexer{input = Input},
    read_char(Lexer).

next_token(#lexer{ch = Ch} = Lexer) ->
    Token = case Ch of
                ${$=} -> equal;
                ${$+} -> plus;
                ${$(} -> lparen;
                ${$)} -> rparen;
                ${${} -> lsquirly;
                ${$}} -> rsquirly;
                ${$,} -> comma;
                ${$;} -> semicolon;
                0     -> eof;
                _     -> illegal
            end,
    {Token, read_char(Lexer#lexer{ch = Token})}.

read_char(#lexer{read_position = RP, input = Input} = Lexer) ->
    if
        RP >= length(Input) ->
            Lexer#lexer{ch = 0};
        true ->
            {Ch, _} = lists:split_at(RP, Input),
            Lexer#lexer{ch = Ch, read_position = RP + 1}
    end

File: csharp/lexer/Lexer.cs
namespace lexer;

public struct Lexer
{
    private int _position;
    private int _read_position;
    private char _current_char;
    private readonly ReadOnlyMemory<char> _input;

    private ReadOnlySpan<char> _inputSpan { get { return _input.Span; } }

    public Lexer(string input)
    {
        _position = 0;
        _read_position = 0;
        _current_char = '\0';
        _input = input.AsMemory();

        ReadCharacter();
    }

    private void ReadCharacter()
    {
        if(_read_position >= _input.Length)
        {
            _current_char = '\0';
        }
        else
        {
            _current_char = _inputSpan[_read_position];
        }

        _position = _read_position;
        _read_position++;
    }

    private void SkipWhitespace()
    {
        while(char.IsWhiteSpace(_current_char))
        {
            ReadCharacter();
        }
    }

    private ReadOnlySpan<char> ReadIdent()
    {
        var start = _position;

        while(char.IsLetter(_current_char) || _inputSpan[_read_position] == '_')
        {
            ReadCharacter();
        }

        return _inputSpan[start.._position];
    }

    private ReadOnlySpan<char> ReadString()
    {
        if(_current_char == '\"')
        {
            ReadCharacter();
        }

        var start = _position;

        while(_current_char != '\0' && _current_char != '\"')
        {
            ReadCharacter();
        }

        if(_current_char == '\0')
        {
            throw new Exception("Unterminated string literal.");
        }

        var stringValue = _inputSpan[start.._position];

        ReadCharacter();

        return stringValue;
    }

    private int ReadInt()
    {
        var start = _position;

        while(char.IsDigit(_current_char))
        {
            ReadCharacter();
        }

        return int.Parse(_inputSpan[start.._position]);
    }

    public TokenInfo NextToken()
    {
        SkipWhitespace();

        if(_current_char == '\"')
        {
            return new TokenInfo(Token.String, ReadString().ToString());
        }

        if (char.IsLetter(_current_char) || _current_char == '_')
        {
            var ident = ReadIdent();
            return ident switch
            {
                "fn" => new TokenInfo(Token.Function),
                "let" => new TokenInfo(Token.Let),
                _ => new TokenInfo(Token.Ident, ident.ToString())
            };
        }

        if(char.IsDigit(_current_char))
        {
            return new TokenInfo(Token.Integer, ReadInt());
        }

        var token = _current_char switch
        {
            '=' => new TokenInfo(Token.Equal),
            '+' => new TokenInfo(Token.Plus),
            ',' => new TokenInfo(Token.Comma),
            ';' => new TokenInfo(Token.Semicolon),
            '(' => new TokenInfo(Token.LParen),
            ')' => new TokenInfo(Token.RParen),
            '{' => new TokenInfo(Token.LSquirly),
            '}' => new TokenInfo(Token.RSquirly),
            '\0' => new TokenInfo(Token.Eof),
            _ => new TokenInfo(Token.Illegal, _current_char)
        };

        ReadCharacter();

        return token;
    }

    public IEnumerable<TokenInfo> ParseTokens()
    {
        TokenInfo tok;
        while((tok = NextToken()).Type != Token.Eof)
        {
            yield return tok;
        }

        yield return tok;
    }
}